### PR TITLE
PYTHON-3728 Simplify convert_codec_options signature

### PR DESCRIPTION
**Pull request by ShaneHarvey**
https://jira.mongodb.org/browse/PYTHON-3728

I'm convinced the Coverity issue is a false positive and that simplifying the convert_codec_options signature should help.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -263,12 +263,7 @@ def build_extension(self, ext):
     Extension(
         "pymongo._cmessage",
         include_dirs=["bson"],
-        sources=[
-            "pymongo/_cmessagemodule.c",
-            "bson/_cbsonmodule.c",
-            "bson/time64.c",
-            "bson/buffer.c",
-        ],
```
Comment (ShaneHarvey) [inline setup.py:271]: I reverted this change made in PYTHON-3717 because I suspect it was unneeded.

File: bson/_cbsonmodule.c [hunk 2]
```diff
@@ -599,8 +599,7 @@ int cbson_convert_type_registry(PyObject* registry_obj, type_registry_t* registr
  * Return 1 on success. options->document_class is a new reference.
  * Return 0 on failure.
  */
-int convert_codec_options(PyObject* self, PyObject* options_obj, void* p) {
-    codec_options_t* options = (codec_options_t*)p;
+int convert_codec_options(PyObject* self, PyObject* options_obj, codec_options_t* options) {
```
Comment (NoahStapp) [inline bson/_cbsonmodule.c:602]: To clarify--what was the reason for the `options` parameter being a generic void pointer before this change?

File: setup.py (hunk 1 above)
Comment (blink1073) [reply setup.py:271]: Looks like it is: `ImportError: /data/mci/881d33b0b5dec4eb9cb9c1f9be10aa61/src/pymongo/_cmessage.cpython-310-x86_64-linux-gnu.so: undefined symbol: cbson_long_long_to_str`

Comment (blink1073) [reply setup.py:271]: Unless removing `extern` is what caused the error.

Comment (ShaneHarvey) [reply setup.py]: I'm going to merge the codec options change and open a new ticket to investigate this issue.

**Comment by ShaneHarvey** [comment]
Ready for another look.

**Review by blink1073 — APPROVED** [review]
LGTM!

**Pull request merged**